src/Data/Singletons/Deriving/Infer.hs

```diff
@@ -13,16 +15,146 @@
+-- context @(Ord a, Ord b)@, since @Left a@ has one field of type @a@, and
+-- @Right b@ has one field of type @b@.
+--
+-- This algorithm is a crude approximation of what GHC actually does when
+-- deriving instances. It is crude in the sense that one can end up with
+-- redundant constraints. For instance, if the data type for which an 'Ord'
+-- instance is being derived is @data Foo = MkFoo Bool Foo@, then the
+-- inferred constraints would be @(Ord Bool, Ord Foo)@. Technically, neither
+-- constraint is necessary, but it is not simple in general to eliminate
+-- redundant constraints like these, so we do not attept to do so. (This is
+-- one reason why @singletons@ requires the use of the @UndecidableInstances@
+-- GHC extension.)
+--
+-- Observant readers will notice that the phrase \"perhaps afer renaming the
+-- type variables\" was casually dropped in step 1 of the above algorithm.
+-- For more information on what this means, refer to the documentation for
+-- infer_ct below.
+inferConstraints :: forall q. DsMonad q => DPred -> DType -> [DCon] -> q DCxt
+inferConstraints pr inst_ty = fmap (nubBy geq) . concatMapM infer_ct
   where
-    infer_ct (DCon _ _ _ fields _) = map (pr `DAppPr`) (tysOfConFields fields)
+    -- A thorny situation arises when attempting to infer an instance context
+    -- for a GADT. Consider the following example:
+    --
+    --   newtype Bar a where
+    --     MkBar :: b -> Bar b
+    --   deriving Show
+    --
+    -- If we blindly apply 'Show' to the field type of @MkBar@, we will end up
+    -- with a derived instance of:
+    --
+    --   instance Show b => Show (Bar a)
+    --
+    -- This is completely wrong, since the type variable @b@ is never used in
+    -- the instance head! This reveals that we need a slightly more nuanced
+    -- strategy for gathering constraints for GADT constructors. To account
+    -- for this, when gathering @field_tys@ (from step 1 in the above algorithm)
+    -- we perform the following extra steps:
+    --
+    -- 1(a). Take the return type of @con@ and unify it with @inst_ty@ (e.g.,
+    --       unify @Bar b@ with @Bar a@). Doing so will produce a substitution
+    --       that maps the universally quantified type variables in the GADT
+    --       (i.e., @b@) to the corresponding type variables in the data type
+    --       constructor (i.e., @a@).
+    -- 1(b). Use the resulting substitution to rename the universally
+    --       quantified type variables of @con@ as necessary.
+    --
+    -- After this renaming, the algorithm will produce an instance context of
+    -- @Show a@ (since @b@ was renamed to @a@), as expected.
+    infer_ct :: DCon -> q DCxt
+    infer_ct (DCon _ _ _ fields mb_res_ty) = do
+      let field_tys = tysOfConFields fields
+      field_tys' <- case mb_res_ty of
+                      Nothing -> pure field_tys
+                      Just res_ty -> do subst <- unify res_ty inst_ty
```

Comment:
You don't really want unification here, as you don't want to map variables that appear in the `inst_ty`. You want matching. And that's what the brand-new `Language.Haskell.TH.Desugar.Subst.matchTy` function does. Use that here (after expanding synonyms, as you've already done).
 
`matchTy` takes a first argument telling it whether or not to ignore kind signatures. Here, you can (`YesIgnore`), because GHC ensures that the result type is well-kinded. You don't have to worry about it.